flax/core/lift.py

```diff
@@ -623,6 +635,129 @@ def mapped(variable_groups, rng_groups, args):
       name='vmap')
 
 
+def _is_axes_leaf(entry):
+  if isinstance(entry, dict) and jax.tree_util.all_leaves(entry.values()):
+    return True
+  # NOTE: `None`s are not considered leaves by `all_leaves`
+  if (isinstance(entry, (tuple, list)) and
+      jax.tree_util.all_leaves(v for v in entry if v is not None)):
+    return True
+  return False
+
+
+def normalize_axes_spec(axes):
+  """Convert to dict representation for an axes spec."""
+  if isinstance(axes, (tuple, list)):
+    axes = {idx: val
+            for idx, val in enumerate(axes)
+            if val not in (Ellipsis, None)}
+  return axes
+
+
+def xmap(
+    fn: Callable[..., Any],
+    variable_axes: Mapping[CollectionFilter, AxesSpec],
+    split_rngs: Mapping[PRNGSequenceFilter, Sequence[str]],
+    in_axes: Any,  # pytree of axis specs
+    out_axes: Any,  # pytree of axis specs
+    axis_sizes: Optional[Mapping[str, int]] = None,
+    axis_resources: Dict[AxisName, ResourceSet] = {},
```

Comment:
nit: we normally use Mapping ot make sure we don't accidentally update user provided mutable dicts